storage/couchdb/couchdb.go

```diff
@@ -0,0 +1,157 @@
+package couchdb
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/docker/go-connections/tlsconfig"
+	"github.com/flimzy/kivik"
+	couch "github.com/go-kivik/couchdb"
+	"github.com/go-kivik/couchdb/chttp"
+	"github.com/sirupsen/logrus"
+)
+
+// ID of a document in CouchDB; includes Rev for updating an existing
+// document
+type ID struct {
+	ID  string `json:"_id,omitempty"`
+	Rev string `json:"_rev,omitempty"`
+}
+
+// Timing can be embedded into other modesl to add time tracking fields
+type Timing struct {
+	CreatedAt time.Time `json:"created_at"`
+	UpdatedAt time.Time `json:"updated_at"`
+	DeletedAt time.Time `json:"deleted_at"`
+}
+
+func createConnection(tlsOpts tlsconfig.Options, host string, username, password string) (*kivik.Client, error) {
+	// remove username and password from URL to avoid authentication step
+	// before we can set the TLS transport
+	_host, err := url.Parse(host)
+	if err != nil {
+		return nil, err
+	}
+	_host.User = nil
+
+	logrus.Debugf("attempting to connect %s to host %s", username, _host.String())
+
+	client, err := kivik.New(context.TODO(), "couch", _host.String())
+	if err != nil {
+		return nil, err
+	}
+
+	t, err := tlsconfig.Client(tlsOpts)
+	if err != nil {
+		return nil, err
+	}
+	// We have to use the original CipherSuites to be able to create a
```

Comment:
It's been a while... I remember having connectivity problems to either CouchDB or Cloudant, not sure. But somehow the CipherSuites were not correct or missing entries that the other side could accept.

Reply:
I added documentation to the top of this script that talks about this.

Reply:
When I disable the line

'''
       // We have to use the original CipherSuites to be able to create a
        // connection with CouchDB
        // t.CipherSuites = tls.Config{}.CipherSuites
'''

I get the following error in the logs:
```
server_1  | {"level":"fatal","msg":"Error starting couchdb driver: Get https://cdb-01.cdb:6984/_session: remote error: tls: insufficient security level","time":"2018-05-21T16:31:44Z"}
```

Reply:
Server side reports this here:

```
SSL: hello: tls_handshake.erl:116:Fatal error: insufficient security
```

On the golang side:

tls.Config{}.CipherSuites seems to set the following:

			TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,
			TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,
			TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
			TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
			TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
			TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,

(non-hardware case from https://golang.org/src/crypto/tls/common.go?s=12277:20617#L912)

whereas we would otherwise only get:

	tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
	tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
	tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
	tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,

(form https://github.com/docker/go-connections/blob/master/tlsconfig/config_client_ciphers.go)

so the difference seems to be:

			TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,
			TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,

The failure seems to be caused by the server, which is written in erlang. It may be related to some erlang regression:

https://stackoverflow.com/questions/36784391/rabbitmq-3-6-1-erlang-18-3-tls-insufficient-security-failures

Reply:
Thanks for clarifying.  We should probably add the CHACHA20 ciphers to the list of supported ciphers in go-connections since they are supported for TLS 1.3 - cc @endophage.